Language functions

louisgosschalk opened the pull request:
Minimum en maximum

louisgosschalk committed:
Minimum en Maximum

Minimum en Maximum begin, heb datavalue integers nodig.

louisgosschalk committed:
Merge branch 'master' into language-functions

louisgosschalk committed:
minimum

changed output of minimum to list of datarows

louisgosschalk committed:
datavalue verwerking aanpassen

louisgosschalk committed:
Minimum & maximum werkend

Deze versie zou een functionerende implementatie van minimum en maximum
moeten zijn

louisgosschalk committed:
minimum maximum kleine wijzigingen

louisgosschalk committed:
MinimumTest

Uitgebreide MinimumTest, test voor minimum of het waarden van string,
int en float goed afhandelt.

louisgosschalk committed:
Maximum aangepast zodat hij int en float goed afhandelt

louisgosschalk committed:
Test voor maximum

louisgosschalk committed:
Merge branch 'master' into language-functions

Boudewijn26 commented:
Ik heb al aangegeven hoe je het beter zou kunnen structureren. In de tussentijd sluit ik de request. Als je vragen hebt, zit ik op telegram.

louisgosschalk committed:
Merge branch 'master' into language-functions

louisgosschalk committed:
Restructure functions

Functions are now structured with abstracts
Will try to make intcompare and floatcompare generally defined

louisgosschalk committed:
Removed debug lines

some system printlines removed, used for debugging

louisgosschalk committed:
rewritten tests

Used databuilder for making tests
much clean
very wow

louisgosschalk committed:
Merge branch 'master' into language-functions

louisgosschalk committed:
more abstractism

abstractified comparison

louisgosschalk committed:
maximum had een foutje

louisgosschalk committed:
Onnodigheid uit maximum en minimum verwijderd

louisgosschalk committed:
edited some tests

Some tests involving the input data are now invalid because the input
folder has been removed.

This is not my work, not sure who made it. there’s no @author

louisgosschalk committed:
Modified function name for compatibility

louisgosschalk committed:
Javadoc

JensLangerak commented:
Ik begrijp uit de comments dat je met Boudewijn hebt overlegd hoe je dit handig kunt doen. Maar dit is denk ik niet hoe Boudewijn het bedoelde. 

Je hebt inderdaad een abstracte class function nodig die door alle functies worden geextend. (eventueel weer subclasses voor familie van functies, maar het wordt vanzelf wel duidelijk of dat nodig is) De class function moet echter alleen defineren wat een functie als input verwacht en wat het als output geeft en verder moet het een abstracte function calculate (o.i.d.) bevatten.
Subclasses definieren vervolgens die calculate functie. Op die manier kun je eenvoudig nieuwe functies toevoegen.
Dus je moet de structuur zodanig hebben, dat wanneer je een nieuwe functie toevoegt, dat je dan de class function niet hoeft aan te passen.

Verder het verschil tussen int en float zou ik opvangen door een generic function. De functies floatCompare en intCompare verschillen nemalijk maar op 2 regels en het verschil is dan dat bij de floatCompare dan float staat en bij intCompare staat er int.

JensLangerak commented:
Omdat ik dus denk dat de structuur nog steeds niet klopt, sluit ik de pull request. Mocht het nog steeds niet duidelijk zijn, dan kun je het vragen via telegram of deze pull request (je kunt gewoon commenten op een gesloten request)

louisgosschalk commented:
ik ben zeer nieuwsgierig naar hoe je het dan anders zou doen. 
Ik heb bewust calculate in de abstracte klasse gezet omdat deze voor min en max exact hetzelfde is behalve voor de vergelijking waar gekeken wordt of het nieuwe element nou groter of kleiner wordt.

louisgosschalk commented:
Ik heb bewust de calculate in de abstracte klasse staan omdat ik anders in zowel minimum als maximum exact dezeflde code heb staan. Dit lijkt mij de enige legitieme oplossing? Voor wat floatCompare en intCompare betreft, heb je zeker gelijk. Dit stond ook nog op mijn to-do lijstje..

louisgosschalk commented:
Ondertussen heb ik trouwens zitten zoeken en heb ik gemerkt dat het lastiger is dan dat het lijkt om aritmetische expressies generiek uit te voeren. In andere woorden; het loopt vast zodra je de generieke values van de columns (float of int) vergelijkt om te bepalen of ze groter of kleiner zijn (en zo het min/maximum bepalen). suggesties worden gewaardeerd

louisgosschalk committed:
Minimum en Maximum ondergebracht in "family"

Abstracte klasse bevat nu alleen input en output model zoals men dat
graag ziet.
Na redelijk onderzoek en een stackoverflow question het besluit genomen
dat aritmetische operaties niet tot nauwelijks generiek uit te voeren
zijn. Daarom laat ik compare in 2 aparte functies.

louisgosschalk committed:
Function

louisgosschalk committed:
Tests aangepast

louisgosschalk committed:
Compare is generiek

Ja, ja, heren. het is gelukt.
Wel inclusief een aantal warnings maar dat is jammer genoeg niet te
voorkomen met generieke aritmetische functies, naar het schijnt.

JensLangerak commented:
maar je moet de actie min en max niet definieren in RowFunction, daar heb je de min en max classes voor. Wat je kunt doen, is aan Minimum en Maximum een functie toevoegen die aangeeft wat er mee met gebeuren. bv de functie genaamd check. Als het resultaat moet worden vervangen, dan geeft de functie check 1 terug, moet die worden toegevoegd, dan geeft hij 0 terug en anders -1. Dus de regel  if((comparison > 0 && minimum) || (comparison < 0 && maximum)){ wordt dan: 
 if(check(comparison) > 0){, de functie check wordt abstract gedefinieerd in RowFuntion en de classes Max en Min implementeren hem.

JensLangerak commented on src/main/java/model/data/process/functions/Function.java:
```diff
@@ -0,0 +1,23 @@
+	
+	public Function(DataTable model, DataDescriber<NumberValue> argument){}
+	
+//	public abstract List<DataRow> calculate();
```
deze moet je niet uitcommenten, er moet een bepaald methode zijn, die je op alle functies kan uitvoeren, anders moet je gaan casten naar verschillende functies en dat is meestal niet de bedoeling

louisgosschalk replied:
Het probleem hierbij is dat average bijvoorbeeld een enkele datavalue returnt terwijl minimum en maximum meerdere datarows returnen.
Daarom stel ik de return type dus niet vast in de abstracte klasse.

JensLangerak commented on src/main/java/model/data/process/functions/RowFunction.java:
```diff
@@ -0,0 +1,83 @@
+	}
+	
+    public List<DataRow> compare() {
+		for(int i = 0; i<table.getRowCount(); i++){
```
Bug: in regel 47 zeg je row = table.getRow(0). Je wilt nu kijken of er rijen groter (/kleiner) of als deze rij de grootste is of er rijen gelijk aan deze rij zijn. Maar de eerste rij die je met rij0 vergelijkt is rij0. rij0 is even groot als rij0 en wordt dus aan het resultaat toegevoegd. Dus nu staat rij0 2 keer in het resultaat.

louisgosschalk replied:
Klopt inderdaad, ik had eerst in de for loop i=1 maar dat is ergens in het omschrijven veranderd. goed gezien

JensLangerak commented on src/main/java/model/data/process/functions/RowFunction.java:
```diff
@@ -0,0 +1,83 @@
+		if(table.getRowCount() == 0)
+			return rowlist; 
+			
+		if(table.getRowCount() == 1) {
```
niet fout, maar wel overbodig. Stel je haalt dit weg, dan zet regel 47 het resultaat op regel0. De conditie voor de for-loop is meteen false (mits de bug wordt opgelost), dus dan wordt alsnog row0 als resultaat terug gegeven.

JensLangerak commented:
verder return je een lege lijst bij een foutieve invoer(bv strings), wil je niet liever een exceptie teruggeven?

louisgosschalk committed:
Aanpassingen n.a.v. comments Jens

Gooit nu een exception bij verkeerde input
bug in loop gefixt

louisgosschalk committed:
Tests aangepast i.v.m. error throwing

louisgosschalk committed:
Speciale exception gemaakt

want dat is leuk

louisgosschalk commented:
Deze commits passen een aantal commentaar punten op, ik pas de code nog aan zodat de calculate ook voldoet

JensLangerak commented:
Ik blijf van mening dat de samenhang van RowData, Minimum en Maximum niet goed is. Maar ik weet niet of ik het duidelijker via de comments kan uitleggen dan ik al heb geprobeerd.
anders hebben we het hier maandag wel over?

louisgosschalk committed:
Zo hoort hij te kloppen

louisgosschalk committed:
kleine revisie

louisgosschalk commented:
Ik heb volgens mij met al je commentaar rekening gehouden en volgens mij hou ik het met deze structuur voldoende rekening met stijlregels.
Nou heb ik ondertussen volgens mijn urenlogboekje ruim 15 uur besteed aan deze opdracht terwijl ik dinsdagochtend functioneel al klaar was met minimum en maximum en ik de rest van die uren dus heb moeten besteden aan herstructureren... Dit lijkt me inderdaad een discussiepunt voor maandag

louisgosschalk committed:
The family class for a function returning a single value

This class is added to specify the return type of functions which
return single values

louisgosschalk committed:
Average calculation

louisgosschalk committed:
javadoc & redundant imports

louisgosschalk committed:
Average test

JensLangerak commented:
Na je laatste wijzigingen denk ik dat hij goed is. ik heb nog een klein dingetje, in een eerdere versie gaf je indien 2 of meer rijen de maximale waarde hadden, alle die rijen terug. Nu niet, is dat bewust gedaan?

louisgosschalk committed:
removed comments and debug lines

JensLangerak commented on src/main/java/model/data/process/functions/Average.java:
```diff
@@ -0,0 +1,72 @@
+			
+			System.out.println("new total " + total);
+		}
+		System.out.println("total after forloop "+total);
```
zijn de prints, achtergebleven van debugging? Zo niet, dan is het mooier als je hier de logger gebruikt

louisgosschalk commented:
Als het goed is worden alle lines met de maximale/minimale value geretourneerd. Dat zou nog steeds zo moeten zijn en volgens mijn tests is dat ook nog steeds zo. Waarom denk je van niet?

louisgosschalk committed:
Extra tests for triple max/minimum results

JensLangerak commented:
je hebt gelijk, sorry ik keek te snel, dacht dat je deze regel ook had weggehaald, maar die staat er nog
else if(comparison == 0)
rowlist.add(compare);

JensLangerak commented:
Dan denk ik dat ie goed is

pjvanwijk commented on src/main/java/model/data/process/functions/RowFunction.java:
```diff
@@ -0,0 +1,75 @@
+		this.table = model;
+		this.argument = argument;
+		this.rowlist = new ArrayList<DataRow>();
+		this.row = null;
```
Initializen met null is eigenlijk overbodig
